File: kotlin/src/main/kotlin/FilesToDateRangeDirectory.kt
package org.example

import org.example.utils.FTDRFileInfo
import org.example.utils.createDuplicateFileInfoGroupsByHash
import java.io.File
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import kotlin.io.path.exists
import kotlin.io.path.getLastModifiedTime

// TODO: check runCatching for every function

fun getDateTimeFormatter(): Result<DateTimeFormatter> = runCatching {
  DateTimeFormatter.ofPattern("yyyy-MM-dd")
}

val addDirectory = fun(directories: MutableList<File>, file: File) {
  directories.add(file)
}

fun isValidDateRangeDirectoryName(name: String): Boolean {
  val spacedHyphen = " - "

  val parseDate = fun(rawDate: String): Result<LocalDate> = runCatching {
    LocalDate.parse(rawDate, getDateTimeFormatter().getOrThrow())
  }

  if (name.contains(spacedHyphen)) {
    val nameParts = name.split(spacedHyphen)
    val firstDate = parseDate(nameParts.first()).getOrElse { return false }
    val secondDate = parseDate(nameParts[1]).getOrElse { return false }
    if (ChronoUnit.DAYS.between(firstDate, secondDate) >= 1) {
      return true
    }
  } else {
    parseDate(name).onSuccess { return true }
  }

  return false
}

fun categorizeFilesAndDirectories(
  destinationDirectory: File
): Pair<MutableList<FTDRFileInfo>, Pair<MutableSet<File>, MutableList<File>>> {
  val files = mutableListOf<FTDRFileInfo>()
  val goodDirectories = mutableSetOf<File>()
  val badDirectories = mutableListOf<File>()

  val categorizeInDirectory = fun(directories: MutableList<File>, file: File) {
    if (isValidDateRangeDirectoryName(file.name)) {
      goodDirectories.add(file)
    } else {
      directories.add(file)
    }
  }

  destinationDirectory.walk().maxDepth(1).forEach { file ->
    categorize(file, files, badDirectories, categorizeInDirectory)
  }

  val directories: Collection<File> = goodDirectories + badDirectories

  directories.forEach { directory ->
    directory.walk().drop(1).forEach { file ->
      categorize(file, files, badDirectories, addDirectory)
    }
  }

  return Pair(files, Pair(goodDirectories, badDirectories))
}

fun categorize(
  file: File,
  files: MutableList<FTDRFileInfo>,
  badDirectories: MutableList<File>,
  handler: (MutableList<File>, File) -> Unit
) = runCatching {
  if (file.isDirectory) {
    handler(badDirectories, file)
  } else if (file.isFile) {
    val size = file.length()
    if (size > 0L) {
      val absolutePath = file.toPath().toAbsolutePath()
      files.add(FTDRFileInfo(
        file = file,
        size = size,
        absolutePath = absolutePath,
        timeModified = absolutePath.getLastModifiedTime().toInstant(),
        newName = null
      ))
    } else {
      // TODO: exception
    }
  } else {
    // TODO: exception
  }
}

fun createHandlers(
  destinationDirectory: File
): List<(MutableList<FTDRFileInfo>, MutableList<File>) -> MutableList<FTDRFileInfo>> {
  val addAllFilesInfo = fun(files: MutableList<File>, filesInfo: List<FTDRFileInfo>) {
    filesInfo.forEach { file ->
      files.add(file.file)
    }
  }

  val addBadFilesInfoAndReplaceGoodFiles = fun(
    badFiles: MutableList<File>, goodFiles: MutableList<FTDRFileInfo>, file: FTDRFileInfo
  ) {
    addAllFilesInfo(badFiles, goodFiles)
    goodFiles.clear()
    goodFiles.add(file)
  }

  val categorizeOnShortestFileNameLength = fun(
    files: MutableList<FTDRFileInfo>, badFiles: MutableList<File>
  ): MutableList<FTDRFileInfo> {
    val good = mutableListOf(files.first())
    var minimumLength = files.first().file.name.length

    files.drop(1).forEach { file ->
      if (file.file.name.length < minimumLength) {
        minimumLength = file.file.name.length
        addBadFilesInfoAndReplaceGoodFiles(badFiles, good, file)
      } else if (file.file.name.length == minimumLength) {
        good.add(file)
      } else {
        badFiles.add(file.file)
      }
    }

    return good
  }

  val categorizeOnValidDateRangeDirectoryName = fun(
    files: MutableList<FTDRFileInfo>, badFiles: MutableList<File>
  ): MutableList<FTDRFileInfo> {
    val tempGood1Files = mutableListOf<FTDRFileInfo>()
    val tempGood2Files = mutableListOf<FTDRFileInfo>()
    val tempBadFiles = mutableListOf<FTDRFileInfo>()

    files.forEach { file ->
      if (file.file.parentFile.parentFile == destinationDirectory) {
        if (isValidDateRangeDirectoryName(file.file.parentFile.name)) {
          tempGood2Files.add(file)
        } else {
          tempGood1Files.add(file)
        }
      } else {
        tempBadFiles.add(file)
      }
    }

    if (tempGood2Files.isNotEmpty()) {
      addAllFilesInfo(badFiles, tempGood1Files)
      addAllFilesInfo(badFiles, tempBadFiles)
      return tempGood2Files
    }

    if (tempGood1Files.isNotEmpty()) {
      addAllFilesInfo(badFiles, tempBadFiles)
      return tempGood1Files
    }

    return tempBadFiles
  }

  val categorizeOnNewestTimeModified = fun(
    files: MutableList<FTDRFileInfo>, badFiles: MutableList<File>
  ): MutableList<FTDRFileInfo> {
    val good = mutableListOf(files.first())
    var newest = files.first().timeModified

    files.drop(1).forEach { file ->
      if (file.timeModified > newest) {
        newest = file.timeModified
        addBadFilesInfoAndReplaceGoodFiles(badFiles, good, file)
      } else if (file.timeModified == newest) {
        good.add(file)
      } else {
        badFiles.add(file.file)
      }
    }

    return good
  }

  val categorizeOnFirstFile = fun(
    files: MutableList<FTDRFileInfo>, badFiles: MutableList<File>
  ): MutableList<FTDRFileInfo> {
    val good = mutableListOf(files.first())

    addAllFilesInfo(badFiles, files.drop(1))

    return good
  }

  return listOf(
    categorizeOnShortestFileNameLength,
    categorizeOnValidDateRangeDirectoryName,
    categorizeOnNewestTimeModified,
    categorizeOnFirstFile
  )
}

fun deleteDuplicateFiles(
  files: MutableList<FTDRFileInfo>, destinationDirectory: File
): Result<Unit> = runCatching {
  val groups = createDuplicateFileInfoGroupsByHash(files, false).getOrThrow() ?: return@runCatching
  val handlers = createHandlers(destinationDirectory)
  val badFiles = mutableListOf<File>()

  files.clear()

  groups.forEachIndexed { index, group ->
    for (handler in handlers) {
      // group and groups[index] are different references
      if (groups[index].size > 1) {
        groups[index] = handler(group, badFiles)
      } else {
        files.add(groups[index].first())
        break
      }
    }
  }

  badFiles.forEach { file ->
    file.delete()
  }
}

fun moveFilesAndFilterGoodDirectories(
  files: MutableList<FTDRFileInfo>, goodDirectories: MutableSet<File>, destinationDirectory: File
) = runCatching {
  if (files.isEmpty()) {
    return@runCatching
  }

  files.sortBy { it.timeModified }

  var fileNames = mutableSetOf<String>()
  var group = mutableListOf<FTDRFileInfo>()

  val replaceFileNamesAndGroup = fun(file: FTDRFileInfo) {
    fileNames = mutableSetOf(file.file.name)
    group = mutableListOf(file)
  }

  replaceFileNamesAndGroup(files.first())

  val formatTimeModified = fun(file: FTDRFileInfo): Result<String> = runCatching {
    file.timeModified.atZone(ZoneId.systemDefault())
      .toLocalDate()
      .format(getDateTimeFormatter().getOrThrow())
  }

  // TODO: can Files.move fail, is so add runCatching
  val moveFilesToDirectory = fun() {
    val firstFile = group.first()
    val lastFile = group.last()

    var directoryName = formatTimeModified(firstFile).getOrThrow()
    if (ChronoUnit.DAYS.between(firstFile.timeModified, lastFile.timeModified) >= 1) {
      directoryName += " - ${formatTimeModified(lastFile).getOrThrow()}"
    }

    val joinedDirectory = File(destinationDirectory.absolutePath, directoryName)
    if (joinedDirectory in goodDirectories) {
      goodDirectories.remove(joinedDirectory)
    } else {
      joinedDirectory.mkdir()
    }

    group.forEach { file ->
      if (file.newName == null) {
        val filePath = Paths.get(joinedDirectory.toString(), file.file.name)
        if (filePath != file.absolutePath) {
          Files.move(file.absolutePath, filePath)
        }
      } else {
        Files.move(file.absolutePath, Paths.get(joinedDirectory.toString(), file.newName))
      }
    }
  }

  files.withIndex().drop(1).forEach { (index, file) ->
    val lastFile = group.last()
    if (ChronoUnit.DAYS.between(lastFile.timeModified, file.timeModified) in 0..3) {
      // TODO: search on ${

      if (file.file.name in fileNames) {
        var disambiguationNumber = 2
        while (disambiguationNumber <= 9) {
          val name = "$file.file.nameWithoutExtension ${disambiguationNumber}.$file.file.extension"
          file.newName = name
          if (name !in fileNames) {
            fileNames.add(name)
            break
          }
          disambiguationNumber++
        }
        if (disambiguationNumber == 9) {
          // TODO: exception
        }
      } else {
        fileNames.add(file.file.name)
      }
      group.add(file)
    } else {
      moveFilesToDirectory()
      replaceFileNamesAndGroup(file)
    }
  }

  if (group.size > 0) {
    moveFilesToDirectory()
  }
}

fun filesToDateRangeDirectory(
  uniqueAbsolutePaths: Array<Path>, destinationDirectory: File
): Result<Unit> = runCatching {
  if (!destinationDirectory.exists()) {
    return@runCatching
  }

  val pair = categorizeFilesAndDirectories(destinationDirectory)
  val files = pair.first
  val goodDirectories = pair.second.first
  val badDirectories = pair.second.second

  uniqueAbsolutePaths.forEach { path ->
    if (path.exists()) {
      categorize(path.toFile(), files, badDirectories, addDirectory)
    }
  }

  deleteDuplicateFiles(files, destinationDirectory).getOrThrow()
  moveFilesAndFilterGoodDirectories(files, goodDirectories, destinationDirectory)

  // There is no need to check if the directory exists before attempting removal.
  badDirectories.asReversed().forEach { directory ->
    directory.delete()
  }
  goodDirectories.forEach { directory ->
    directory.delete()
  }
}

File: go/files_to_date_range_directory.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/markuskreukniet/markus-tools/go/utils"
)

const dateLayout = "2006-01-02" // YYYY-MM-DD

func isValidDateRangeDirectoryName(name string) bool {
	const spacedHyphen = " - "

	parseDate := func(rawDate string) (time.Time, error) {
		date, err := time.Parse(dateLayout, rawDate)
		if err != nil {
			return time.Time{}, err
		}
		return date, nil
	}

	if strings.Contains(name, spacedHyphen) {
		nameParts := strings.Split(name, spacedHyphen)
		firstDate, err := parseDate(nameParts[0])
		if err != nil {
			return false
		}
		secondDate, err := parseDate(nameParts[1])
		if err != nil {
			return false
		}
		daysDifference := secondDate.Sub(firstDate).Hours() / 24
		if daysDifference >= 1 {
			return true
		}
	} else {
		if _, err := parseDate(name); err == nil {
			return true
		}
	}
	return false
}

// TODO: naming
type dateRangeArg struct {
	directoryName string
	filePath      string
}

func addDirectory(directories *[]string, arg dateRangeArg) {
	*directories = append(*directories, arg.filePath)
}

func categorizeFilesAndDirectories(destinationDirectory string) ([]utils.DateRangeFileInfo, map[string]struct{}, []string, error) {
	var files []utils.DateRangeFileInfo
	goodDirectoryPaths := make(map[string]struct{})
	var badDirectoryPaths []string

	categorizeInDirectory := func(directoryPaths *[]string, arg dateRangeArg) {
		if isValidDateRangeDirectoryName(arg.directoryName) {
			goodDirectoryPaths[arg.filePath] = struct{}{}
		} else {
			*directoryPaths = append(*directoryPaths, arg.filePath)
		}
	}

	entries, err := os.ReadDir(destinationDirectory)
	if err != nil {
		return nil, nil, nil, err
	}

	for _, entry := range entries {
		info, err := entry.Info()
		if err != nil {
			return nil, nil, nil, err
		}
		categorize(
			info, filepath.Join(destinationDirectory, entry.Name()), &files, &badDirectoryPaths, categorizeInDirectory,
		)
	}

	directories := append([]string{}, badDirectoryPaths...)

	for path := range goodDirectoryPaths {
		directories = append(directories, path)
	}

	for _, path := range directories {
		err := filepath.Walk(path, func(filePath string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}

			if path != filePath {
				categorize(info, filePath, &files, &badDirectoryPaths, addDirectory)
			}

			return nil
		})
		if err != nil {
			return nil, nil, nil, err
		}
	}

	return files, goodDirectoryPaths, badDirectoryPaths, nil
}

func categorize(
	info os.FileInfo,
	filePath string,
	files *[]utils.DateRangeFileInfo,
	badDirectoryPaths *[]string,
	handler func(*[]string, dateRangeArg),
) error {
	name := filepath.Base(filePath)

	if info.IsDir() {
		handler(badDirectoryPaths, dateRangeArg{
			directoryName: name,
			filePath:      filePath,
		})
	} else if info.Mode().IsRegular() {
		size := info.Size()
		if size > 0 {
			*files = append(*files, utils.DateRangeFileInfo{
				Size:         size,
				Path:         filePath,
				Name:         name,
				TimeModified: info.ModTime(),
			})
		} else {
			// TODO: error
		}
	} else {
		// TODO: error
	}

	return nil
}

func createHandlers(
	destinationDirectory string) []func([]utils.DateRangeFileInfo, *[]utils.DateRangeFileInfo,
) []utils.DateRangeFileInfo {
	appendBadFilesAndReplaceGoodFiles := func(
		badFiles *[]utils.DateRangeFileInfo, goodFiles *[]utils.DateRangeFileInfo, file utils.DateRangeFileInfo,
	) {
		*badFiles = append(*badFiles, *goodFiles...)
		*goodFiles = []utils.DateRangeFileInfo{file}
	}

	categorizeOnShortestFileNameLength := func(
		files []utils.DateRangeFileInfo, badFiles *[]utils.DateRangeFileInfo,
	) []utils.DateRangeFileInfo {
		getNameLength := func(file utils.DateRangeFileInfo) int {
			return len(file.Name)
		}

		good := []utils.DateRangeFileInfo{files[0]}
		var minimumLength = getNameLength(files[0])

		for _, file := range files[1:] {
			nameLength := getNameLength(file)
			if nameLength < minimumLength {
				minimumLength = nameLength
				appendBadFilesAndReplaceGoodFiles(badFiles, &good, file)
			} else if nameLength == minimumLength {
				good = append(good, file)
			} else {
				*badFiles = append(*badFiles, file)
			}
		}

		return good
	}

	categorizeOnValidDateRangeDirectoryName := func(
		files []utils.DateRangeFileInfo, badFiles *[]utils.DateRangeFileInfo,
	) []utils.DateRangeFileInfo {
		var tempGood1Files []utils.DateRangeFileInfo
		var tempGood2Files []utils.DateRangeFileInfo
		var tempBadFiles []utils.DateRangeFileInfo

		for _, file := range files {
			directoryPath := filepath.Dir(file.Path)
			if filepath.Dir(directoryPath) == destinationDirectory {
				if isValidDateRangeDirectoryName(directoryPath) {
					tempGood2Files = append(tempGood2Files, file)
				} else {
					tempGood1Files = append(tempGood1Files, file)
				}
			} else {
				tempBadFiles = append(tempBadFiles, file)
			}
		}

		if len(tempGood2Files) > 0 {
			*badFiles = append(*badFiles, tempGood1Files...)
			*badFiles = append(*badFiles, tempBadFiles...)
			return tempGood2Files
		}

		if len(tempGood1Files) > 0 {
			*badFiles = append(*badFiles, tempBadFiles...)
			return tempGood1Files
		}

		return tempBadFiles
	}

	categorizeOnNewestTimeModified := func(
		files []utils.DateRangeFileInfo, badFiles *[]utils.DateRangeFileInfo,
	) []utils.DateRangeFileInfo {
		good := []utils.DateRangeFileInfo{files[0]}
		newest := files[0].TimeModified

		for _, file := range files[1:] {
			if file.TimeModified.After(newest) {
				newest = file.TimeModified
				appendBadFilesAndReplaceGoodFiles(badFiles, &good, file)
			} else if file.TimeModified.Equal(newest) {
				good = append(good, file)
			} else {
				*badFiles = append(*badFiles, file)
			}
		}

		return good
	}

	categorizeOnFirstFile := func(
		files []utils.DateRangeFileInfo, badFiles *[]utils.DateRangeFileInfo,
	) []utils.DateRangeFileInfo {
		good := []utils.DateRangeFileInfo{files[0]}

		*badFiles = append(*badFiles, files[1:]...)

		return good
	}

	return []func([]utils.DateRangeFileInfo, *[]utils.DateRangeFileInfo) []utils.DateRangeFileInfo{
		categorizeOnShortestFileNameLength,
		categorizeOnValidDateRangeDirectoryName,
		categorizeOnNewestTimeModified,
		categorizeOnFirstFile,
	}
}

func deleteDuplicateFiles(files *[]utils.DateRangeFileInfo, destinationDirectory string) error {
	groups, err := utils.CreateDuplicateFileInfoGroupsByHash(*files, false)
	if err != nil {
		return err
	}

	handlers := createHandlers(destinationDirectory)
	var badFiles []utils.DateRangeFileInfo

	*files = (*files)[:0]

	for index, group := range groups {
		for _, handler := range handlers {
			// group and groups[index] are different references
			if len(groups[index]) > 1 {
				groups[index] = handler(group, &badFiles)
			} else {
				*files = append(*files, groups[index][0])
				break
			}
		}
	}

	for _, file := range badFiles {
		if err := os.Remove(file.GetPath()); err != nil {
			return err
		}
	}

	return nil
}

// TODO: rename destinationDirectory to destinationDirectoryPath on other places
func moveFilesAndFilterGoodDirectories(
	files []utils.DateRangeFileInfo, goodDirectoryPaths *map[string]struct{}, destinationDirectoryPath string,
) error {
	if len(files) == 0 {
		return nil
	}

	sort.Slice(files, func(i, j int) bool {
		return files[i].TimeModified.Before(files[j].TimeModified)
	})

	// TODO: also do this set logic in Kotlin
	// A set provides O(1) access time but is unordered, so we cannot reliably retrieve the first value, for example.
	// To maintain order, we also use a slice alongside the set.
	var fileNames map[string]struct{}
	var group []utils.DateRangeFileInfo

	replaceFileNamesAndGroup := func(file utils.DateRangeFileInfo) {
		fileNames = map[string]struct{}{file.Name: {}}
		group = []utils.DateRangeFileInfo{file}
	}

	formatTimeModified := func(file utils.DateRangeFileInfo) string {
		return file.TimeModified.Format(dateLayout)
	}

	moveFilesToDirectory := func() error {
		firstFile := group[0]
		lastFile := group[len(group)-1]

		directoryName := formatTimeModified(firstFile)
		if lastFile.TimeModified.Sub(firstFile.TimeModified).Hours() >= 24 {
			directoryName += fmt.Sprintf(" - %s", formatTimeModified(lastFile))
		}

		joinedDirectoryPath := filepath.Join(destinationDirectoryPath, directoryName)

		if _, exists := (*goodDirectoryPaths)[joinedDirectoryPath]; exists {
			delete(*goodDirectoryPaths, joinedDirectoryPath)
		} else {
			if err := utils.CreateDirectory(joinedDirectoryPath); err != nil {
				return err
			}
		}

		for _, file := range group {
			joinedFilePath := filepath.Join(joinedDirectoryPath, file.Name)
			if joinedFilePath != file.Path {
				if err := os.Rename(file.Path, joinedFilePath); err != nil {
					return err
				}
			}
		}

		return nil
	}

	replaceFileNamesAndGroup(files[0])

	// TODO: also fix in Kotlin code

	for i := 1; i < len(files); i++ {
		lastFile := group[len(group)-1]
		if files[i].TimeModified.Sub(lastFile.TimeModified).Hours() <= 72 {
			if _, exists := fileNames[files[i].Name]; exists {
				extension := filepath.Ext(files[i].Name)
				nameWithoutExtension := strings.TrimSuffix(files[i].Name, extension)

				// TODO: disambiguationNumber should start at 0 or 1?
				disambiguationNumber := 2
				for ; disambiguationNumber <= 9; disambiguationNumber++ {
					name := fmt.Sprintf("%s %d%s", nameWithoutExtension, disambiguationNumber, extension)
					if _, exists := fileNames[name]; !exists {
						files[i].Name = name
						break
					}
				}
				if disambiguationNumber == 10 {
					// TODO: error
				}
			}
			fileNames[files[i].Name] = struct{}{}
			group = append(group, files[i])
		} else {
			if err := moveFilesToDirectory(); err != nil {
				return err
			}
			replaceFileNamesAndGroup(files[i])
		}
	}

	if len(group) > 0 {
		if err := moveFilesToDirectory(); err != nil {
			return err
		}
	}

	return nil
}

func filesToDateRangeDirectory(uniqueFileSystemNodes []utils.FileSystemNode, destinationDirectory string) error {
	files, goodDirectoryPaths, badDirectoryPaths, err := categorizeFilesAndDirectories(destinationDirectory)
	if err != nil {
		return err
	}

	for _, node := range uniqueFileSystemNodes {
		info, err := os.Stat(node.Path)
		if err != nil {
			return err
		}
		categorize(info, node.Path, &files, &badDirectoryPaths, addDirectory)
	}

	if err := deleteDuplicateFiles(&files, destinationDirectory); err != nil {
		return err
	}

	if err := moveFilesAndFilterGoodDirectories(files, &goodDirectoryPaths, destinationDirectory); err != nil {
		return err
	}

	// Remove the bad empty directories
	// There is no need to check if the directory exists before attempting removal.
	for i := len(badDirectoryPaths) - 1; i >= 0; i-- {
		if err := os.Remove(badDirectoryPaths[i]); err != nil {
			return err
		}
	}

	for path := range goodDirectoryPaths {
		if err := os.Remove(path); err != nil {
			return err
		}
	}

	return nil
}
